collector/kernel/kernel_collector.cc

```diff
@@ -545,6 +550,18 @@ void KernelCollector::on_error(int error)
   stop_all_timers();
 }
 
+void KernelCollector::restart()
+{
+  /* we don't want attempts to perform IO on the channel */
+  heartbeat_sender_.stop();
+
+  // flush the channel so previously sent messages make it to the pipeline server
+  upstream_connection_.flush();
+
+  // close the channel, it will trigger reconnect via KernelCollector::Callbacks::on_closed() which calls enter_try_connecting()
+  upstream_connection_.close();
+}
```

Comment:
This is where the KernelCollector is self-restarted.  It follows the same steps that occur when the KernelCollector is restarted due to a broken connection with the pipeline server.